### PR TITLE
Complete reorg of tutorials and guides

### PR DESCRIPTION
**Pull request by melissawm**
Hello,

This PR is the last in a series to close out #24632 .

In particular, it implements the following:

- Splitting the Advanced Autodiff section into subsections
- Adding a debugging section to the quickstart
- Move the discussion on jitting class methods to Sharp Bits.

In addition, I also did some reorganization of links and anchors to match the new layout. 

A few notes:
- The "Custom derivative rules" from the old advanced autodiff doc was almost identical to the Custom derivative rules notebook, so I tried updating the notebook to the latest version of the content in the old section.
- I moved the "Is JAX faster than NumPy?" section in the FAQ to an H2, since I think this was meant to be a standalone question in the FAQ. Let me know if that's not right.
- I came up with a different title to the Higher order derivatives section/doc and those need a review.
- The new debugging section in the quickstart is very short but I think it's useful. Happy to address any feedback there.

With this, I believe all items in the initial reorg proposal are addressed.

Thanks!

### CODE REVIEW

File: docs/gradient-checkpointing.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ kernelspec:
 
 In this tutorial, you will learn how to control JAX automatic differentiation's saved values using {func}`jax.checkpoint` (also known as {func}`jax.remat`), which can be particularly helpful in machine learning.
 
-If you are new to automatic differentiation (autodiff) or need to refresh your memory, JAX has {ref}`automatic-differentiation` and {ref}`advanced-autodiff` tutorials.
+If you are new to automatic differentiation (autodiff) or need to refresh your memory, JAX has an {ref}`automatic-differentiation` tutorials and several {ref}`Advanced automatic differentiation guides <advanced_guides>`.
```
Comment (emilyfertig) [inline docs/gradient-checkpointing.md]: typo: s/tutorials/tutorial

File: docs/notebooks/thinking_in_jax.md [hunk 2]
```diff
@@ -474,6 +474,46 @@ For more on pseudo random numbers in JAX, see the [Pseudorandom numbers tutorial
 
 +++
 
+## Debugging
+
+Debugging JAX code can be challenging due to its functional programming model and the fact that JAX code is often transformed via JIT compilation or vectorization. However, JAX provides several tools to help with debugging.
+
+- For simple inspection, use [`jax.debug.print`](https://docs.jax.dev/en/latest/_autosummary/jax.debug.print.html). Unlike Python's built-in `print`, which executed at trace-time, before runtime values exist, `jax.debug.print` will print the actual traced (dynamic) array values during JIT-compiled execution.
+- For logging, use [`jax.debug.callback`](https://docs.jax.dev/en/latest/_autosummary/jax.debug.callback.html) which enables you to pass in a Python function that can be called inside of a staged JAX program.
```
Comment (emilyfertig) [inline docs/notebooks/thinking_in_jax.md]: Instead of `jax.debug.callback`, could you add a bullet point and a quick demo of `jax.debug_nans`, and replace the `jax.debug.callback` demo with `jax.debug.print` (maybe showing how Python `print` vs. `debug.print` behave differently with jit)? 

`debug.print` and `debug_nans` are more basic and widely-used I think than `debug.callback` so they'd fit better in the quickstart.

File: docs/advanced_guides.rst [hunk 3]
```diff
@@ -32,9 +32,16 @@ operations.
    :maxdepth: 1
 
    notebooks/autodiff_cookbook
-   notebooks/Custom_derivative_rules_for_Python_code
    notebooks/autodiff_remat
-   advanced-autodiff
+
```
Comment (emilyfertig) [inline docs/advanced_guides.rst]: Could you make it so that the Automatic differentiation/Advanced AD sections expand under a dropdown menu (titled "Automatic differentiation"), like the current "Introduction to debugging" and "Debugging runtime values" sections? Since there's 6 of them all on AD, I think that'd be nice to group them.

**Review by emilyfertig — CHANGES_REQUESTED** [review]
Thank you!

File: docs/advanced_guides.rst (hunk 3 above)
Comment (melissawm) [reply docs/advanced_guides.rst]: Will do!

File: docs/notebooks/thinking_in_jax.md (hunk 2 above)
Comment (melissawm) [reply docs/notebooks/thinking_in_jax.md]: Let me know if this is better! I didn't want to duplicate too much content but added a quick example.

File: docs/advanced_guides.rst (hunk 3 above)
Comment (melissawm) [reply docs/advanced_guides.rst]: Let me know what you think of this organization. I also had to fix the debugging ToC since I noticed the docs were appearing twice in the main ToC.

**Comment by emilyfertig** [comment]
Thanks, this looks great! Could you squash the commits? (I didn't realize I needed to add the "documentation" label to get the docs to render, so that should re-trigger once you push again. I just want to look at the rendered ToC before I approve.)

**Comment by melissawm** [comment]
Done, thanks!

**Review by emilyfertig — APPROVED** [review]
